Precompute IDFs in scoring process

drien opened the pull request:
Thanks for the library!

I've been prototyping something with this and found the performance on my dataset (~1200 documents, average of 240 words) wasn't quite enough for how I'd been hoping to use it, but when I profiled the code I saw that most of the time was actually being spent in the `getIDF` function, which happened to be called redundantly in the main scoring loop.

Precomputing the IDF values for each keyword like this is a pretty big performance boost, I'm seeing 80%+ faster searches on my data, down from 3.6s to 280ms with three keywords.

drien committed:
Precompute IDFs

FurkanToprak commented on BM25.ts:
```diff
@@ -59,10 +59,15 @@ export default function BM25(
+    obj[keyword] = getIDF(keyword, documents);
+    return obj;
+  }, {});
+
```
Let's pass in a Map object instead of an arbitrary Object. This way we can type check for indexes that result in undefined values.

For example

  const idfByKeyword = keywords.reduce((idfMap, keyword) => {
    idfMap.set(keyword, getIDF(keyword, documents));
    return idfMap;
  }, new Map<type, type>());

drien replied:
Cool, done.

FurkanToprak reviewed (changes requested):
Hey @drien, thanks for the contribution! Contributors like you are the backbone of this library :)

I've added some feedback but otherwise I'm happy to merge this if you add some additional testing or provide some more thorough benchmarking justifying this change.

drien committed:
Use set for precomputed IDF, add types

drien commented:
Here are some benchmarks searching against the [20 Newsgroups dataset](http://qwone.com/~jason/20Newsgroups/). 99.94% faster on a two word query over 20k documents!

I haven't added tests because testing based on performance benchmarks is not especially reliable IME, though you'll see in the script I added assertions that the output between this update and the original is identical.

`Original` meaning the library as installed off of npm, `Updated` being the compiled version from this branch.

Version | Number of documents | Number of words in the query | Run time
```
Original 	 500 docs 	 1 words 	 313.2352420091629ms
Updated 	 500 docs 	 1 words 	 13.642560988664627ms
Original 	 5000 docs 	 1 words 	 34165.26628398895ms
Updated 	 5000 docs 	 1 words 	 144.3623880147934ms
Original 	 19997 docs 	 1 words 	 375871.29902899265ms
Updated 	 19997 docs 	 1 words 	 419.90233102440834ms

Original 	 500 docs 	 2 words 	 636.7892889976501ms
Updated 	 500 docs 	 2 words 	 16.285903990268707ms
Original 	 5000 docs 	 2 words 	 70638.97426700592ms
Updated 	 5000 docs 	 2 words 	 159.61680600047112ms
Original 	 19997 docs 	 2 words 	 864217.5051890016ms
Updated 	 19997 docs 	 2 words 	 457.79560601711273ms

Original 	 500 docs 	 3 words 	 824.1660939753056ms
Updated 	 500 docs 	 3 words 	 13.979169011116028ms
```

Here's the quick script I threw together to generate that (I cut it off as it was starting to take too long with longer queries):
```javascript
const BM25 = require("okapibm25").default;
const BM25Test = require("./dist/BM25.js").default;

var assert = require("assert");

const { glob } = require("glob");
const fs = require("fs");

const { performance } = require("node:perf_hooks");

const main = async () => {
  const files = await glob("20_newsgroups/**/*");

  const docs = [];

  for (let file of files) {
    if (fs.lstatSync(file).isDirectory()) {
      continue;
    }
    let data = fs.readFileSync(file, "utf8");

    if (data) {
      docs.push(data.toLowerCase());
    }
  }

  const queries = [
    "Hardware",
    "Encryption algorithms",
    "Performance issues Windows",
    "Baseball World Series results",
    "Comparison Mac PC hardware capabilities",
  ].map((i) => i.toLowerCase().match(/\w+/g));

  for (let query of queries) {
    let start = performance.now();
    let control = BM25(docs.slice(0, 500), query, { k1: 1.3, b: 0.9 });
    console.log(
      `Original \t 500 docs \t ${query.length} words \t ${performance.now() - start}ms`,
    );
    start = performance.now();
    let test = BM25Test(docs.slice(0, 500), query, { k1: 1.3, b: 0.9 });
    console.log(
      `Updated \t 500 docs \t ${query.length} words \t ${performance.now() - start}ms`,
    );
    assert(JSON.stringify(control) === JSON.stringify(test));

    start = performance.now();
    control = BM25(docs.slice(0, 5000), query, { k1: 1.3, b: 0.9 });
    console.log(
      `Original \t 5000 docs \t ${query.length} words \t ${performance.now() - start}ms`,
    );
    start = performance.now();
    test = BM25Test(docs.slice(0, 5000), query, { k1: 1.3, b: 0.9 });
    console.log(
      `Updated \t 5000 docs \t ${query.length} words \t ${performance.now() - start}ms`,
    );
    assert(JSON.stringify(control) === JSON.stringify(test));

    start = performance.now();
    control = BM25(docs, query, { k1: 1.3, b: 0.9 });
    console.log(
      `Original \t ${docs.length} docs \t ${query.length} words \t ${performance.now() - start}ms`,
    );
    start = performance.now();
    test = BM25Test(docs, query, { k1: 1.3, b: 0.9 });
    console.log(
      `Updated \t ${docs.length} docs \t ${query.length} words \t ${performance.now() - start}ms`,
    );
    assert(JSON.stringify(control) === JSON.stringify(test));

    console.log();
  }
};

main();
```

FurkanToprak commented:
Hey, your changes have now been published. Thank you for your contribution!
